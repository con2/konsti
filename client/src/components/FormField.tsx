import React, { FC, ReactElement } from 'react';
import { useTranslation } from 'react-i18next';
import styled from 'styled-components';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { WrappedFieldProps } from 'redux-form';

interface CustomFieldProps {
  type?: string;
}

export const FormField: FC<WrappedFieldProps & CustomFieldProps> = (
  props: WrappedFieldProps & CustomFieldProps
): ReactElement => {
  const { type } = props;
  const { name, onChange, onBlur, onDragStart, onDrop, onFocus } = props.input;
  const { touched, error } = props.meta;
  const { t } = useTranslation();

  const [fieldType, setFieldType] = React.useState<string>('');

  React.useEffect(() => {
    setFieldType(type ?? 'text');
  }, [type]);

  const classNames = ['form-input'];

  if (type === 'checkbox') {
    classNames.push('checkbox');
  }

  const togglePasswordVisibility = (): void => {
    if (fieldType === 'password') setFieldType('text');
    else if (fieldType === 'text') setFieldType('password');
  };

  return (
    <>
      <FormRow>
        <StyledFormField>
          <StyledInput
            className={classNames.join(' ')}
            id={name}
            name={name}
            onBlur={onBlur}
            onChange={onChange}
            onDragStart={onDragStart}
            onDrop={onDrop}
            onFocus={onFocus}
            placeholder={t(name)}
            type={fieldType}
          />
          {type === 'checkbox' && <label htmlFor={name}>{t(name)}</label>}
        </StyledFormField>

        {type === 'password' && (
          <FormFieldIcon>
            <FontAwesomeIcon
              icon={fieldType === 'password' ? 'eye' : 'eye-slash'}
              onClick={togglePasswordVisibility}
            />
          </FormFieldIcon>
        )}
      </FormRow>

      {touched && error && (
        <FormFieldError>
          <FormFieldErrorMessage>{t(error)}</FormFieldErrorMessage>
        </FormFieldError>
      )}
    </>
  );
};

const FormRow = styled.div`
  align-items: center;
  display: flex;
  flex: 0 1 auto;
  flex-direction: row;
  width: 50%;
  justify-content: flex-start;

  @media (max-width: ${(props) => props.theme.breakpointPhone}) {
    width: 100%;
  }
`;

const StyledFormField = styled.div`
  align-items: center;
  display: flex;
  flex: 0 1 auto;
  flex-direction: row;
  padding: 8px 0;
  width: 80%;
`;

const FormFieldIcon = styled.span`
  padding: 0 0 0 8px;
  font-size: ${(props) => props.theme.iconSize};
`;

const FormFieldError = styled.div`
  display: flex;
  background-color: ${(props) => props.theme.backgroundHighlight};
  color: ${(props) => props.theme.error};
  width: 50%;

  @media (max-width: ${(props) => props.theme.breakpointPhone}) {
    width: 100%;
  }
`;

const FormFieldErrorMessage = styled.span`
  padding: 4px 0 4px 10px;
`;

const StyledInput = styled.input`
  &.form-input {
    border: 1px solid ${(props) => props.theme.borderInactive};
    color: ${(props) => props.theme.buttonText};
    height: 34px;
    padding: 0 0 0 10px;
    width: 100%;
  }

  &.checkbox {
    margin-right: 10px;
    width: 16px;
  }
`;
