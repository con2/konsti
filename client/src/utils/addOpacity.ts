export const addOpacity = (rgb: string, opacity: string): string => {
  const formattedRgb = rgb.replace("rgb", "").replace("(", "").replace(")", "");
  const rgba = `rgba(${formattedRgb}, ${opacity})`;
  return rgba;
};
