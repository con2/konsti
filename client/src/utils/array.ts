export const moveArrayItem = <T>(
  arr: readonly T[],
  from: number,
  to: number
): readonly T[] => {
  return arr.reduce((prev: T[], current, index, self) => {
    if (from === to) {
      prev.push(current);
    }
    if (index === from) {
      return prev;
    }
    if (from < to) {
      prev.push(current);
    }
    if (index === to) {
      prev.push(self[from]);
    }
    if (from > to) {
      prev.push(current);
    }
    return prev;
  }, []);
};

export const insertByIndex = <T>(
  arr: readonly T[],
  newItem: T,
  insertAt: number
): readonly T[] => [...arr.slice(0, insertAt), newItem, ...arr.slice(insertAt)];
