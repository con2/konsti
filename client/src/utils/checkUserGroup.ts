import { UserGroup } from "shared/types/models/user";

export const isUser = (userGroup: UserGroup): boolean => {
  return userGroup === UserGroup.USER;
};

export const isAdmin = (userGroup: UserGroup): boolean => {
  return userGroup === UserGroup.ADMIN;
};

export const isAdminOrHelp = (userGroup: UserGroup): boolean => {
  return userGroup === UserGroup.ADMIN || userGroup === UserGroup.HELP;
};
