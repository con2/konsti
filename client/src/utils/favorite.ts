import { Game } from 'typings/game.typings';
import { submitUpdateFavorites } from 'views/my-games/myGamesActions';
import { Dispatch } from 'redux';

export interface UpdateFavoriteOpts {
  game: Game;
  action: string;
  favoritedGames: readonly Game[];
  username: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  dispatch: Dispatch<any>;
}

export const updateFavorite = async (
  updateFavoriteOpts: UpdateFavoriteOpts
): Promise<void> => {
  const {
    game,
    action,
    favoritedGames,
    username,
    dispatch,
  } = updateFavoriteOpts;

  if (!game || !game.gameId) return;

  const gameIndex = favoritedGames.findIndex(
    (favoritedGame) => favoritedGame.gameId === game.gameId
  );
  const allFavoritedGames = favoritedGames.slice();

  if (action === 'add' && gameIndex === -1) {
    allFavoritedGames.push(game);
  } else if (action === 'del' && gameIndex > -1) {
    allFavoritedGames.splice(gameIndex, 1);
  }

  const favoriteData = {
    username: username,
    favoritedGames: allFavoritedGames,
  };

  try {
    await dispatch(submitUpdateFavorites(favoriteData));
  } catch (error) {
    throw new Error(`submitUpdateFavorites error: ${error}`);
  }
};
