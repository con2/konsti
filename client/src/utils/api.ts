import axios, { AxiosInstance } from 'axios';
import { config } from 'config';
import { getJWT } from 'utils/getJWT';

export const api: AxiosInstance = axios.create({
  baseURL: `${config.apiServerURL}/api`,
  timeout: 60000, // 60s
  headers: {
    'Content-Type': 'application/json',
  },
});

api.interceptors.request.use((config) => {
  const authToken = getJWT();
  if (authToken) {
    config.headers.Authorization = `Bearer ${authToken}`;
  }
  return config;
});
