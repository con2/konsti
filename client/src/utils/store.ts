import {
  createStore,
  combineReducers,
  applyMiddleware,
  CombinedState,
  AnyAction,
} from 'redux';
import thunk from 'redux-thunk';
import { composeWithDevTools } from 'redux-devtools-extension';
import { config } from 'config';
import { loadSession } from 'utils/localStorage';
import { RootState } from 'typings/redux.typings';
import { SUBMIT_LOGOUT } from 'typings/logoutActions.typings';

// Reducers
import { reducer as formReducer } from 'redux-form';
import { allGamesReducer } from 'views/all-games/allGamesReducer';
import { loginReducer } from 'views/login/loginReducer';
import { myGamesReducer } from 'views/my-games/myGamesReducer';
import { signupReducer } from 'views/signup/signupReducer';
import { adminReducer } from 'views/admin/adminReducer';
import { resultsReducer } from 'views/results/resultsReducer';

// Set reducers
export const appReducer = combineReducers({
  form: formReducer,
  allGames: allGamesReducer,
  login: loginReducer,
  myGames: myGamesReducer,
  signup: signupReducer,
  admin: adminReducer,
  results: resultsReducer,
});

// Reducer to reset state
const rootReducer = (
  state: RootState,
  action: AnyAction
): CombinedState<RootState> => {
  if (action.type !== SUBMIT_LOGOUT) return appReducer(state, action);

  const newState = appReducer(undefined, action);
  newState.admin = state.admin;
  newState.allGames = state.allGames;
  return newState;
};

const middlewares = applyMiddleware(thunk);

const composeEnhancers = composeWithDevTools({
  trace: config.reduxTrace,
  traceLimit: 25,
});

const enhancer = composeEnhancers(middlewares);

// Load persisted state from localStorage
const persistedState = loadSession();

// Create a Redux store object that holds the app state
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-expect-error
export const store = createStore(rootReducer, persistedState, enhancer);
