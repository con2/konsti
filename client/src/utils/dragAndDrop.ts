import _ from 'lodash';
import { moveArrayItem, insertByIndex } from 'utils/array';
import { Game, DnDUpdatedPositions, DnDMove } from 'typings/game.typings';

export const reorder = (
  list: readonly Game[],
  startIndex: number,
  endIndex: number
): readonly Game[] => {
  return moveArrayItem(list, startIndex, endIndex);
};

export const move = (
  sourceList: readonly Game[],
  destinationList: readonly Game[],
  sourceMove: DnDMove,
  destinationMove: DnDMove
): DnDUpdatedPositions => {
  return {
    [sourceMove.droppableId]: _.without(
      sourceList,
      sourceList[sourceMove.index]
    ),

    [destinationMove.droppableId]: insertByIndex(
      destinationList,
      sourceList[sourceMove.index],
      destinationMove.index
    ),
  };
};
