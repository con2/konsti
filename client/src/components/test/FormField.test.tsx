import React from 'react';
import { shallow } from 'enzyme';
import { WrappedFieldProps } from 'redux-form';
import { FormField } from '../FormField';
import { store } from 'utils/store';

describe('FormField', () => {
  it('should render correctly', () => {
    const props: WrappedFieldProps = {
      meta: {
        active: false,
        asyncValidating: false,
        autofilled: false,
        dirty: false,
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-expect-error
        dispatch: () => store.dispatch(''),
        error: 'validation.required',
        form: 'registration',
        initial: undefined,
        invalid: true,
        pristine: true,
        submitFailed: true,
        submitting: false,
        touched: true,
        valid: false,
        visited: false,
        warning: undefined,
      },
      input: {
        name: 'password',
        value: '',
        onBlur: () => {},
        onChange: () => {},
        onDragStart: () => {},
        onDrop: () => {},
        onFocus: () => {},
      },
    };
    const component = shallow(<FormField {...props} />);
    expect(component).toMatchSnapshot();
  });
});
