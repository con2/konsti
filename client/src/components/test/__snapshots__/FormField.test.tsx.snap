// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormField should render correctly 1`] = `
<Fragment>
  <FormField__FormRow>
    <FormField__StyledFormField>
      <FormField__StyledInput
        className="form-input"
        id="password"
        name="password"
        onBlur={[Function]}
        onChange={[Function]}
        onDragStart={[Function]}
        onDrop={[Function]}
        onFocus={[Function]}
        placeholder="password"
        type=""
      />
    </FormField__StyledFormField>
  </FormField__FormRow>
  <FormField__FormFieldError>
    <FormField__FormFieldErrorMessage>
      validation.required
    </FormField__FormFieldErrorMessage>
  </FormField__FormFieldError>
</Fragment>
`;
